certification.py

```diff
@@ -203,15 +216,38 @@ class PartyTypeParty(ModelSQL, ModelView):
     valid = fields.Function(fields.Boolean('Valid'), 'is_valid')
 
     def is_valid(self, name):
+        pool = Pool()
+        Date = pool.get('ir.date')
+
+        today = Date.today()
+
         required_documents = set()
-        for line in self.party_type.document_types:
-            if line.required:
-                required_documents.add(line.document_type.id)
+        required_substitute = {}
+        for document_type in self.party_type.document_types:
+            if document_type.required:
+                document_type_id = document_type.document_type.id
+                required_documents.add(document_type_id)
+                if document_type.document_type.substitute:
+                    sub_doc_type_id = document_type.document_type.substitute.id
+                    required_substitute[sub_doc_type_id] = document_type_id
 
         for document in self.party.documents:
+            document_type_id = document.document_type.id
+            if (document.state == 'approved' and
+                    document_type_id in required_documents):
+                # in case has expiration_date, not valid
+                if document.expiration_date and document.expiration_date < today:
+                    continue
+                required_documents.remove(document_type_id)
+
+            # remove in case has approved document is required and has substitute
+            document_type_parent_id = required_substitute.get(document_type_id)
             if (document.state == 'approved' and
-                    document.document_type.id in required_documents):
-                required_documents.remove(document.document_type.id)
+                    document_type_parent_id in required_documents):
+                # in case has expiration_date, not valid
+                if document.expiration_date and document.expiration_date < today:
+                    continue
+                required_documents.remove(document_type_parent_id)
```

Comment:
Add test party_certification and check party valid documents at e059673098d4e1d03f22e2bef62bfab25bac8aae

Reply:
> I'm not aggree.
> 
> In case first document is "valid" (approved), return True and not check other documents.
> 
> An example from document type configuration:
> 
> ```
> document A - required  + substitute document C
> document B - required
> ```
> 
> when generate documents in the party have:
> 
> ```
> document A - required + substitute
> document B - required
> document C - required in case document A is not approved
> ```
> 
> In case you approve "document B", your proposal return True. But see that "document A" or "document C" are required.

I don't think that is possible with my code given that in the loop I only return False when something goes wrong. In order for the method to return True all types must have been checked.

Reply:
6de0ade518597c6c9417fb78454ddcbaa10ab7f1